enhancements/shared-and-guaranteed-cpu-pools.md

```diff
@@ -0,0 +1,407 @@
+(shared vs. guaranteed). To automate this, an admission hook will be added to the 
+[Kubernetes API Server in OpenShift](https://github.com/openshift/kubernetes/tree/master/openshift-kube-apiserver/admission), 
+which will mutate each pod definition as it is created:
+* For containers with guaranteed whole CPUs, add a new resource request/limit `openshift.io/guaranteed-cpus` 
+equal to the number of CPUs requested in millicore units.
+* For containers with non-guaranteed CPU requests (or guaranteed fractional CPU requests) add a new resource 
+request/limit `openshift.io/shared-cpus` equal to the number of CPUs requested in millicore units. Note that 
+one of the restrictions for extended resources is that the request/limit must match - in the case of a container 
+with differing CPU requests/limits, the CPU request value will be used - this will preserve existing scheduler
+behaviour as only the CPU requests are used to choose the node a pod will run on.
+
+Here are the CPU requests/limits from an example container that uses shared CPUs:
+```yaml
+requests:
+  cpu: 200m
+limits:
+  cpu: 400m
+```
+This would be mutated as follows:
+```yaml
+requests:
+  cpu: 200m
+  openshift.io/shared-cpus: 200
+limits:
+  cpu: 400m
+  openshift.io/shared-cpus: 200
+```
+
+Kubelet's cgroup creation/management code will continue to use the CPU requests/limits to configure the CFS shares/quota
+for each container and will not require any modifications.
+
+### Workflow Description
+
+**cluster creator** is a human user responsible for deploying a cluster.
+
+1. The cluster creator creates a Performance Profile for the NTO and specifies a shared CPU partition.
+2. The cluster creator creates a machine config manifest to write a configuration file for kubelet to specify
+the shared and isolated CPUs (TODO: or this could be the config option).
+3. The cluster creator (TODO: does what???) to enable the new admission hook.
+4. The cluster creator then creates the cluster.
+5. The kubelet starts up, finds the configuration file and initializes the `defaultCpuSet` and `guaranteedCpuSet` 
+based on the shared/isolated CPUSets specified in the config file.
+6. The kubelet advertises `openshift.io/shared-cpus` and `openshift.io/guaranteed-cpus` extended resources 
+on the node based on the shared/isolated CPUSets specified in the config file.
+7. Something schedules:
+   * a pod with the `target.workload.openshift.io/management` annotation in a namespace 
+with the `workload.openshift.io/allowed` management annotation. The admission hook ignores the pod as it will
+be handled by the managementcpusoverride admission hook.
+   * a pod with Burstable or BestEffort QoS. The admission hook modifies the pod,
+adding `openshift.io/shared-cpus` requests/limits for each container, matching the CPU requests.
+   * a pod with Guaranteed QoS. The admission hook modifies the pod,
```

Comment:
Just to make it more clear, I would dedicate a separate bullet for guaranteed pod with integral CPUs and guaranteed pod with non integral CPUs.

Reply:
Will do